### DISCUSSION TITLE
Joins with prisma

### DISCUSSION THREAD

**Discussion by m-shariq** [prisma/prisma#19748] [Q&A] [answered]
**Issue**

I am using Prisma with PostgreSQL and expect many joins in my queries, and my app is performance critical. I have seen many issues related to joins that Prisma doesn't support and are relatively slow.

So, my question is the same, does Prisma now support joins ? if not, then which ORM or just query builder is best (performance effective) with type-safety, type-checks and types definitions?

Thanks

**Comment by nikolasburk** [comment, marked as answer]
Hey @m-shariq, thanks a lot for raising this!

It's true that Prisma currently doesn't do JOINs for relational queries. Instead, it sends individual queries and joins the data on the application level. However, this doesn't mean that Prisma's relational queries are per se slower — it's important to understand that the performance of a query very much depends on the exact query that's executed and the specific dataset that's being queried!

I want to elaborate a bit on the tradeoffs of this decision.

The reason Prisma uses this strategy is because in a lot of real-world applications with large datasets, DB-level JOINs can become quite expensive. This has been discussed in this [issue](https://github.com/prisma/prisma/issues/4997#issuecomment-516806095):

> The total cost of executing a complex join is often higher than executing multiple simpler queries. This is why the Prisma query engine currently defaults to multiple simple queries in order to optimise overall throughput of the system. | https://github.com/sorenbs (Founder/CEO of Prisma)

> But Prisma is designed towards generalized best practices, and in the "real world" with huge tables and hundreds of fields, single queries are not the best approach. | https://github.com/guerrerocarlos

Also, this chapter about [Query Performance Optimization](https://www.oreilly.com/library/view/high-performance-mysql/9780596101718/ch04.html) from the [High Performance MySQL](https://learning.oreilly.com/library/view/high-performance-mysql/9780596101718/) book has some great insights. One of the techniques it mentions is JOIN decomposition: 

> Many high-performance web sites use join decomposition. You can decompose a join by running multiple single-table queries instead of a multitable join, and then performing the join in the application. 

Again, Prisma chose this strategy initially to account for real-world scenarios with data-heavy applications where DBL-level JOINs become way too expensive. For smaller datasets, the difference of using a JOIN or sending multiple queries can most often be neglected, so it was important for us to make sure Prisma works well in _all_ cases and that's why we have initially picked this tradeoff for Prisma ORM.

All that being said, there are _of course_ scenarios where JOINs are a lot more performance than sending individual queries. We know this and that's why we are currently working on enabling JOINs in Prisma Client queries as well 😄 You can follow the development on the [roadmap](https://www.notion.so/prismaio/Optimize-Prisma-Client-Query-Performance-68caf089581e4d07ac154dd34ad73639). 

Also, be aware that performance is a high priority for us at Prisma at the moment! We recently [improved the cold start times of Prisma in Serverless environments by 9x](https://www.prisma.io/blog/prisma-and-serverless-73hbgKnZ6t) and are expecting to make similar strides when it comes to query performance in the next few weeks/months.

If you are actually experiencing concrete, slow queries with Prisma, we would be super thankful if you could [open a GitHub issue](https://github.com/prisma/prisma/issues/new/choose) with more details about the query and the dataset so that we can make sure to address it in our effort to improve Prisma Client performance!

If this answers your question, it would be great if you could mark this Discussion as _answered_ to indicate that it has been resolved.

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by nikolasburk** [reply]
Also, FYI: There's a great tool called [SafeQL](https://safeql.dev/compatibility/prisma.html) that improves the developer experience when writing raw SQL queries in Prisma a lot! So if you ever need to optimize a Prisma Client query using `$queryRaw`, you can use that for syntax highlighting and type checking.

**Reply by m-shariq** [reply]
Hi @nikolasburk, thanks for the clarification and sharing the above tools. My question was related to joins in Prisma and I got my answer.
